### PR TITLE
✨ Preact implementation for bento-dailymotion

### PR DESCRIPTION
**Pull request by rebeccanthomas**
This PR includes the initial preact implementation for bento conversion of the amp-dailymotion component.

### CODE REVIEW

File: extensions/amp-dailymotion/1.0/test/validator-amp-dailymotion.html
Comment (Gregable) [inline extensions/amp-dailymotion/1.0/test/validator-amp-dailymotion.html:1]: There should be a matching .out file generated for this test.

**Review by Gregable — CHANGES_REQUESTED** [review]

File: extensions/amp-dailymotion/1.0/test/validator-amp-dailymotion.html
Comment (alanorozco) [reply extensions/amp-dailymotion/1.0/test/validator-amp-dailymotion.html:1]: @rebeccanthomas Since this is the Preact implementation, let's remove AMP-mode related files for now. This includes validation files, and `amp-dailymotion.js`/`base-element.js`.

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.js:121]: It looks like 0.1 shows video controls by default without taking in the prop
```suggestion
      controls={true}
      ref={ref}
```

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.js:44]: `break` is not needed following `return`
```suggestion
```

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.js:78]: This can be either the video API or a defined Dailymotion API:
```suggestion
 * @param {{current: ?VideoWrapperDef.Api}} ref
```
```suggestion
 * @param {{current: ?DailymotionDef.DailymotionApi}} ref
```

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (caroqliu) [inline extensions/amp-dailymotion/dailymotion-api.js:23]: looks like the linter added the extra param because of the casing difference:
```suggestion
 * @param {string} videoId
```

**Review by caroqliu — CHANGES_REQUESTED** [review]
Nice start! Just a few suggestions ⬇️

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/dailymotion-api.js:59]: esLint seems to say that iterator spreading is forbidden in this circumstance, with the code like this: 
`  let iframeUrl = addParamsToUrl(
    `https://www.dailymotion.com/embed/video/${encodeURIComponent(
      videoId
    )}?api=1&html=1&app=amp`,
    dict({
      'endscreen-enable': endscreenEnable,
      'info': info,
      'mute': mute,
      'sharing-enable': sharingEnable,
      'start': start,
      'ui-highlight': uiHighlight,
      'ui-logo': uiLogo,
    }),
    ...implicitParams,
  );`

Is there a different place I should be putting the spread?

Comment (caroqliu) [reply extensions/amp-dailymotion/dailymotion-api.js:59]: Ah, `dict` doesn't like the spread syntax, we will have to use `Object.assign`:
```js
return addParamsToUrl(
    `https://www.dailymotion.com/embed/video/${encodeURIComponent(
      videoId
    )}?api=1&html=1&app=amp`,
    Object.assign(
      dict({
        'endscreen-enable': endscreenEnable,
        'info': info,
        'mute': mute,
        'sharing-enable': sharingEnable,
        'start': start,
        'ui-highlight': uiHighlight,
        'ui-logo': uiLogo,
      }),
      implicitParams
    )
  );
```

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/0.1/amp-dailymotion.js]: Done!

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/dailymotion-api.js]: They do not! good catch thanks!a

File: extensions/amp-dailymotion/1.0/component.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/1.0/component.js]: Is there a reason to use one over the other?

Comment (caroqliu) [reply extensions/amp-dailymotion/1.0/component.js]: Meaning wise, the first one is already defined and can be used if the API is identical. The second one we would use if the API is different from the other and needs to be separately defined. cc/ @samouri in case type checking changes-to-come has another answer here.

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (caroqliu) [inline extensions/amp-dailymotion/dailymotion-api.js:60]: I think we want to reverse the logic here because it should always be muted if autoplay is enabled, even if `data-mute` gives a different value via the API. We see this because in 0.1 the original `mute` param value gets overwritten under the `autoplay && isAutoplaySupported` conditions.
```suggestion
         autoplay ? 1 : mute,
```

File: extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js
Comment (caroqliu) [inline extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js:92]: Should this be `1` for consistency?

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.js:28]: This can be removed as unused

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.js:115]: need to propagate `autoplay` here to get it to work. please add storybook knobs for all these props so we can verify behavior
```suggestion
      autoplay={autoplay}
      controls={true}
```

File: extensions/amp-dailymotion/1.0/test/test-component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/test/test-component.js:49]: Can we add another test case where `mute="false"` and all other `false` prop values are provided via omission? The assertions should yield the same results if our default values are in place.

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (caroqliu) [inline extensions/amp-dailymotion/dailymotion-api.js:19]: Looking at `vimeo-api.js`, it looks like only 0.1 checks `isAutoplaySupported` whereas 1.0 does not. Likely this is because it relies on `win` to be defined, which it won't be on Preact's first render.

**Review by caroqliu — CHANGES_REQUESTED** [review]

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (caroqliu) [reply extensions/amp-dailymotion/dailymotion-api.js:60]: Whoops, you're right. 😆  Thank you!

File: extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js:92]: I changed it to true because we are now directly passing the result of a boolean statement into the URL.

Comment (rebeccanthomas) [reply extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js:92]: Ah! I saw your suggestion for how to fix that! Applying now :)

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.js:68]: We need `playing` event to trigger the autoplaying UI
```suggestion
    dispatchCustomEvent(currentTarget, VideoEvents.PLAYING);
```

Comment (caroqliu) [reply extensions/amp-dailymotion/1.0/component.js:68]: It may be helpful to have a `DAILYMOTION_EVENTS` to do this conversion for both versions via the dailymotion-api, similar to `VIMEO_EVENTS` and `BRIGHTCOVE_EVENTS`

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [reply extensions/amp-dailymotion/1.0/component.js:127]: Btw you can provide `true` via presence
```suggestion
      controls
```

**Review by Gregable — APPROVED** [review]
The validator files have been removed. I don't think you need my review anymore, but in case it's important for github, you have my approval for 0 files in this CL. :)

File: extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js
Comment (caroqliu) [inline extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js:57]: FYI the way that I've seen test files annotate is:
```js
      const dailymotion = await getDailymotion('x2m8jpp', /* optResponsive */ true);
```

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (caroqliu) [inline extensions/amp-dailymotion/dailymotion-api.js:59]: This can be reused in 0.1's `sendCommand_` fn

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (caroqliu) [inline extensions/amp-dailymotion/dailymotion-api.js:56]: I noticed all applications of this helper in 0.1 and 1.0 only ever pass one boolean, if any. Is the array really necessary? Can we update calls to only pass the one optional param?
```suggestion
 * @param {?boolean} param
```

File: extensions/amp-dailymotion/1.0/component.type.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.type.js:14]: These should all be `|undefined`

File: extensions/amp-dailymotion/1.0/test/test-component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/test/test-component.js:47]: nit: you can make a variable if you find yourself referencing the same prop over and over
```suggestion
    const src = iframe.prop('src');
    expect(src).to.contain('endscreen-enable=false');
    expect(src).to.contain('sharing-enable=false');
    expect(src).to.contain('ui-highlight=444444');
    expect(src).to.contain('info=false');
```

File: extensions/amp-dailymotion/1.0/component.js
Comment (caroqliu) [inline extensions/amp-dailymotion/1.0/component.js:47]: FYI you can use `redispatch` here 
```suggestion
redispatch(currentTarget, event, {
    [DailymotionEvents.API_READY]: 'canplay',
    [DailymotionEvents.END]: [VideoEvents.ENDED, VideoEvents.PAUSE],
    [DailymotionEvents.PAUSE]: VideoEvents.PAUSE,
    [DailymotionEvents.PLAY]: VideoEvents.PLAYING,
  });
```

**Review by caroqliu — COMMENTED** [review]
Almost there. Is this error expected?
Screen Shot 2021-09-01 at 4 26 12 PM

File: extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/0.1/test/test-amp-dailymotion.js:57]: Ah Thanks! That is way more readable!

File: extensions/amp-dailymotion/1.0/component.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/1.0/component.js:47]: It seems like`redispatch` only works in an amp environment but i've update the `dispatchCustomEvent` calls to be individual calls instead of sending the array.

**Review by rsimha — APPROVED** [review]

**Comment by caroqliu** [comment]
> Almost there. Is this error expected?
> Screen Shot 2021-09-01 at 4 26 12 PM

It looks like this error is a result of `VideoIframe` applying `DEFAULT_SANDBOX` whereas 0.1 does not apply any `sandbox`. @alanorozco is it desirable for `1.0` to break in this way? Looks like it yields the above error in the console but not seeing anything different in behavior.

**Review by caroqliu — APPROVED** [review]
Approved once events are `redispatch`'d as discussed in prior comments. Error message can be resolved in a separate PR if preferred -- please file a tracking issue to keep track of these types of tasks.

**Comment by alanorozco** [comment]
@caroqliu It's fine to leave the `allow-presentation` error as-is if it doesn't break our current functionality. We don't care about [`PresentationSession`](https://developer.mozilla.org/en-US/docs/Web/API/PresentationRequest)

**Review by alanorozco — APPROVED** [review]

**Comment by rebeccanthomas** [comment]
> Approved once events are `redispatch`'d as discussed in prior comments. Error message can be resolved in a separate PR if preferred -- please file a tracking issue to keep track of these types of tasks.

As we discussed offline for the purposes of the preact/react component we will keep using dispatchCustomEvent.

File: extensions/amp-dailymotion/1.0/amp-dailymotion.css
Comment (alanorozco) [inline extensions/amp-dailymotion/1.0/amp-dailymotion.css:13]:
```suggestion
amp-dailymotion:not(.i-amphtml-built) > :not([placeholder]):not([slot="i-amphtml-svc"]) {
```

File: extensions/amp-dailymotion/1.0/component.js
Comment (alanorozco) [inline extensions/amp-dailymotion/1.0/component.js:64]:
```suggestion
    uiLogoDisable,
```

File: extensions/amp-dailymotion/1.0/amp-dailymotion.css
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/1.0/amp-dailymotion.css:1]: This file seems to be required in order for the unit tests to run, regardless of whether there is an associated 1.0/amp-dailymotion.js

Comment (caroqliu) [reply extensions/amp-dailymotion/1.0/amp-dailymotion.css:1]: That's odd. What is the error message without it?

File: extensions/amp-dailymotion/1.0/component.js
Comment (rebeccanthomas) [reply extensions/amp-dailymotion/1.0/component.js:64]: As discussed offline, in the interest of parity with the dailymotion documentation and the 0.1 functionality we are leaving these attrs as they are.

Comment (alanorozco) [reply extensions/amp-dailymotion/1.0/component.js:64]: As we discussed, let's keep these as Dailymotion defines them, where they're true by default. 

Should still be renamed
```suggestion
    uiLogo,
```

File: extensions/amp-dailymotion/0.1/amp-dailymotion.js
Comment (alanorozco) [inline extensions/amp-dailymotion/0.1/amp-dailymotion.js:250]:
```suggestion
      mute === 'true',
```

File: extensions/amp-dailymotion/dailymotion-api.js
Comment (alanorozco) [inline extensions/amp-dailymotion/dailymotion-api.js:50]: nit: If these are real booleans we don't need triple equality.

```suggestion
        // In the dailymotion API endscreenEnable, info, sharingEnable, and uiLogo
        // all default to true, so if the attr is not marked as false do not add it to the URL
        'endscreen-enable': !endscreenEnable ? endscreenEnable : undefined,
        'info': !info ? info : undefined,
        // In order to support autoplay the video needs to be muted on load so we
        // dont receive an unmute event which prevents the video from autoplay.
        'mute': mute || (autoplay && isAutoplaySupported(win)) ? 1 : undefined,
        'sharing-enable': !sharingEnable ? sharingEnable : undefined,
        'start': start,
        'ui-highlight': uiHighlight,
        'ui-logo': !uiLogo ? uiLogo : undefined,
```

**Pull request merged**
